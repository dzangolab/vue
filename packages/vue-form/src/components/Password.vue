<template>
  <div class="field password">
    <label v-if="label" for="password">
      {{ label }}
    </label>
    <Field
      v-slot="{ field, meta }"
      v-bind="modelValue"
      :name="name"
      :rules="fieldSchema"
    >
      <input
        v-bind="field"
        :class="{
          invalid: meta.touched && !meta.valid,
          valid: meta.dirty && meta.valid,
        }"
        tabindex="0"
        type="password"
      />
      <ErrorMessage :name="name" />
    </Field>
  </div>
</template>

<script lang="ts">
export default {
  name: "Password",
};
</script>

<script setup lang="ts">
import { toFieldValidator } from "@vee-validate/zod";
import { ErrorMessage, Field } from "vee-validate";

import { passwordSchema } from "../schemas";

import type { PasswordErrorMessages, StrongPasswordOptions } from "../types";
import type { PropType } from "vue";

const props = defineProps({
  errorMessages: {
    default: () => {
      return {
        required: "A password is required",
        weak: "This password is too weak",
      };
    },
    required: false,
    type: Object as PropType<PasswordErrorMessages>,
  },
  label: {
    default: null,
    required: false,
    type: String as PropType<string | null | undefined>,
  },
  modelValue: {
    default: null,
    required: false,
    type: String as PropType<string | null | undefined>,
  },
  name: {
    default: "password",
    required: false,
    type: String as PropType<string>,
  },
  options: {
    default: () => {
      return {};
    },
    required: false,
    type: Object as PropType<StrongPasswordOptions>,
  },
});

defineEmits(["update:modelValue"]);

const fieldSchema = toFieldValidator(
  passwordSchema(
    props.errorMessages,
    props.options as StrongPasswordOptions & { returnScore: false | undefined }
  )
);
</script>
