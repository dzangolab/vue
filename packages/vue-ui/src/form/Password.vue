<template>
  <div class="field password">
    <label v-if="label" for="password">
      {{ label }}
    </label>
    <Field
      v-slot="{ field, meta }"
      v-bind="modelValue"
      :name="name"
      :rules="fieldSchema"
    >
      <input
        v-bind="field"
        :class="{
          invalid: meta.touched && !meta.valid,
          valid: meta.dirty && meta.valid,
        }"
        tabindex="0"
        type="password"
      />
      <ErrorMessage name="password" />
    </Field>
  </div>
</template>

<script lang="ts">
export default {
  name: "Password",
};
</script>

<script setup lang="ts">
import { toFieldValidator } from "@vee-validate/zod";
import validator from "validator";
import { ErrorMessage, Field } from "vee-validate";
import { z } from "zod";

import type { PasswordErrorMessages } from "../types/form";
import type { StrongPasswordOptions } from "../types/validator";
import type { PropType } from "vue";

const props = defineProps({
  errorMessages: {
    default: () => {
      return {
        required: "Email address is required",
        weak: "This password is too weak",
      };
    },
    required: false,
    type: Object as PropType<PasswordErrorMessages>,
  },
  label: {
    default: "",
    required: false,
    type: String as PropType<string>,
  },
  modelValue: {
    required: true,
    type: String as PropType<string | null | undefined>,
  },
  name: {
    default: "password",
    required: false,
    type: String as PropType<string>,
  },
  options: {
    default: () => {
      return {};
    },
    required: false,
    type: Object as PropType<StrongPasswordOptions>,
  },
});

defineEmits(["update:modelValue"]);

const defaultOptions = {
  minLength: 8,
  minLowercase: 0,
  minUppercase: 0,
  minNumbers: 0,
  minSymbols: 0,
  returnScore: false,
  pointsPerUnique: 1,
  pointsPerRepeat: 0.5,
  pointsForContainingLower: 10,
  pointsForContainingUpper: 10,
  pointsForContainingNumber: 10,
  pointsForContainingSymbol: 10,
};

const options = {
  ...defaultOptions,
  ...props.options,
};

const fieldSchema = toFieldValidator(
  z
    .string({
      required_error: props.errorMessages.required,
    })
    .refine(
      (value): boolean => {
        return validator.isStrongPassword(
          value,
          options as StrongPasswordOptions & { returnScore: false | undefined }
        );
      },
      {
        message: props.errorMessages.weak,
      }
    )
);
</script>
