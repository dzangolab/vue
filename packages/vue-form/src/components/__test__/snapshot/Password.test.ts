import { mount, RouterLinkStub } from "@vue/test-utils";
import { describe, expect, it } from "vitest";

import Password from "../../Password.vue";

describe("Password", () => {
  it("matches snapshot", () => {
    const wrapper = mount(Password, {
      global: {
        stubs: {
          RouterLink: RouterLinkStub,
        },
      },
      props: {
        label: "Password",
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });
});
