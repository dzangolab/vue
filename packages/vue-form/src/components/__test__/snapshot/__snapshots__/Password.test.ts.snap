// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Password > matches snapshot 1`] = `
<div
  class="field password"
>
  <label
    for="password"
  >
    Password
  </label>
  
  <div
    class="password-input"
  >
    <input
      class=""
      name="password"
      placeholder=""
      tabindex="0"
      type="password"
    />
    <span
      class="eye-icon"
    >
      
      <img
        alt="eye-icon"
        src="/src/assets/svg/eye-slash.svg"
      />
      
    </span>
  </div>
  <!---->
  
</div>
`;
